### DISCUSSION TITLE
Build ToaruOS without docker?

### DISCUSSION THREAD

**Discussion by mxlgv** [klange/toaruos#273] [Q&A] [answered]
Is it possible to build a package of tools (gcc, binutils) and ToaruOS itself without using docker? I just can't stand docker). And I find it extremely inconvenient. For example, I have no idea how to build SDL in docker. Or I'm using it incorrectly.

**Comment by klange** [comment, marked as answer]
You can manually install the toolchain, which is how I generally do development. I do not provide instructions for doing this, as the process will vary between different host environments, but you can look at the scripts used to build the docker containers as a reference: https://github.com/klange/toaruos/tree/master/util/docker

You will need to make sure you clone with full submodules to get the patched `binutils` and `gcc` trees. After that, there are essentially two parts to building the toolchain locally:

- Installing necessary packages: The list, at least for older versions of Ubuntu like what the Docker containers have been using, is here: `build-essential python3 xorriso genext2fs mtools gnu-efi git automake autoconf wget libgmp-dev libmpfr-dev libmpc-dev flex bison texinfo dosfstools`
- Building the toolchain: This is building the customized configurations of Binutils and GCC. This is mostly automated by this utility script: https://github.com/klange/toaruos/blob/master/util/build-toolchain.sh

Once you have the toolchain built, you can "activate" it (add its `bin` directory to your `$PATH`) and then try to `./configure --host=x86_64-pc-toaru --target=x86_64-pc-toaru` in an autotools project.

For SDL, if you're building from my fork, that should work okay. For new ports, the first stumbling block is usually a `config.sub` that looks at the target triplet and needs to be updated.

**Reply by mxlgv** [reply]
Thanks for the help!

**Reply by mxlgv** [reply]
@klange, hello! Sorry to bother you again. I managed to assemble the toolchain and it even assembles "helloworld". But when I try to build ToaruOS:
```bash
make base/lib/libc.so && make -j4
```

I am getting an error:
```log
make: *** No rule to build target 'base/usr/lib/libgcc_s.so.1' required for 'base/lib/libtoaru_list.so'. Stop.
```

**Reply by mxlgv** [reply]
I fixed the issue. It turns out that the script cannot be run from the utils folder, it must be from the root.
